### ISSUE TITLE
registry: Accept wildcard is not supported

### ISSUE THREAD

**Issue by rogpeppe** [distribution/distribution#3938]
[RFC 9110](https://www.rfc-editor.org/rfc/rfc9110.html#section-12.5.1) allows wildcard (e.g. `*/*`, `application/*`) to be specified for the Accept header.

However, as of commit 97b1d649c4938d0f608d96454d6a8326b1f96acd, the parsing of the Accept header implemented by the registry [only supports explicit content types](https://github.com/distribution/distribution/blob/97b1d649c4938d0f608d96454d6a8326b1f96acd/registry/handlers/manifests.go#L106-L117).

It would be useful to follow the relevant web standard here and allow wildcard Accept headers so a content-agnostic client can avoid enumerating all possible manifest types, a practice which is both bulky and potentially fragile to maintain over time.

For example, passing the header:

```
Accept: */*
```
to a manifest GET request should allow all possible manifest types.

**Comment by deleteriousEffect** [comment]
@rogpeppe is there an extant content-agnostic client in use that would benefit from this change?

**Comment by rogpeppe** [comment]
@deleteriousEffect [This one](https://github.com/rogpeppe/ociregistry/commit/1d3d9d0c3e04fbcb802dfac4a4e17960a3342ecf) :)

OK, I may be a bit biased there, as not many other folks are writing an arbitrary registry transformer like I am, but in all seriousness, I did a brief survey and every single client tool I could find just enumerates all the possible media manifest types without actually being more specific as to the one that they're requesting, which seems to me to indicate that they don't care that much exactly what they're getting.

Examples:
- [skopeo](https://github.com/containers/skopeo/blob/68c76d70f87077fdebd693a19e7d74d85f0c7a2e/vendor/github.com/containers/image/v5/docker/docker_image.go#L143-L150)
- [crane](https://github.com/google/go-containerregistry/blob/c3f8a49229d3ccaa5096ff65e1696ce129c1091e/pkg/v1/remote/descriptor.go#L81)
- [oras-go](https://github.com/oras-project/oras-go/blob/0e20275bbbd71f4619d9c1b9da150c4f586dff3b/registry/remote/manifest.go#L54-L59)

Although the set of manifest types is [configurable](https://pkg.go.dev/oras.land/oras-go/v2/registry/remote#Repository) in `oras-go`, a quick code search persuades me that no-one is actually using that configurability.

And the spec [specifies](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#pulling-manifests) that clients should check that the media type matches what they expect anyway, so there's not a great deal of difference in practice: clients can discard the received manifest content if its content type doesn't match what is required.

If a new manifest type is created, none of these tools will be able to deal with the new manifest type, even though in principle I believe they could do so in many circumstances (for example when displaying manifest content for a user).

**Comment by deleteriousEffect** [comment]
@rogpeppe I'm hesitant to support this, there are really only a handful of manifest media types and these media types can be _quite_ distinct from another. For example, the list in your application does not include the signed and unsigned versions of docker schema 1 manifests the wildcard would also include those as well. These old manifest types (in my personal opinion) should really no longer be supported by anything, so the utility of the wildcard is already a little dubious.

**Comment by jonjohnsonjr** [comment]
> For example, the list in your application does not include the signed and unsigned versions of docker schema 1 manifests the wildcard would also include those as well. These old manifest types (in my personal opinion) should really no longer be supported by anything, so the utility of the wildcard is already a little dubious.

It's a funny example to use, because if you _do_ try to use a wildcard Accept with distribution, the registry will _downconvert_ everything to schema 1, which is incredibly surprising!

FWIW, I made GCR respect `Accept: */*` after about a dozen customer issues where folks tried to use `curl` with the registry and were very surprised that they were getting back schema 1 images. I did not have any complaints for years after that because it's the behavior that you would expect. Having to do this non-standard `Accept` header negotiation is super confusing, and I think most people agree that the current behavior is a mistake: https://github.com/opencontainers/distribution-spec/issues/212#issuecomment-726127431

Basically the whole point of the `Accept` header was to differentiate between very old clients (that assumed everything was schema 1 and didn't understand mediaTypes) and modern clients that looked at the Content-Type header. We no longer care about those very old clients, so it doesn't make much sense to cater to them.

**Comment by Jamstah** [comment]
Doesn't the registry do on the fly conversion to the type requested where possible? You're not querying to find the type, your telling the registry what type you want.

**Comment by jonjohnsonjr** [comment]
> Doesn't the registry do on the fly conversion to the type requested where possible?

Yes, and it's (IMO) very wrong/surprising thing for the registry to do given the content-addressable nature of the system. That behavior existed for very old clients that didn't send `Accept` headers at all.

https://docs.docker.com/registry/spec/manifest-v2-2/#backward-compatibility

**Comment by deleteriousEffect** [comment]
> Yes, and it's (IMO) very wrong/surprising thing for the registry to do given the content-addressable nature of the system. 

We don't downconvert when manifests are being retrieved by their content address: https://github.com/distribution/distribution/blob/main/registry/handlers/manifests.go#L177

It is surprising behavior now, but not as surprising as breaking the majority of clients from one release to the next would have been at the time. We'd need a major version bump to change this behavior now, either through stopping downconversion or dropping support for schema 1 completely. The current maintainers didn't make this choice, but we have to live with the consequences. We have to be careful about not breaking existing clients and workflows, unfortunately even at the expense of new clients and workflows.

> FWIW, I made GCR respect Accept: */* after about a dozen customer issues where folks tried to use curl with the registry and were very surprised that they were getting back schema 1 images.

Passing no Accept headers is still different from passing wildcard accept headers, they'd still get back schema 1 images without passing some kind of accept headers, right?

**Comment by mxmeinhold** [comment]
As of a buildkit upgrade from v0.25.2 to v0.31.2, I'm finding that it's sending a much narrower accept header, (I think that should be [mediatype plus `*/*`](https://github.com/containerd/containerd/blob/v2.2.2/core/remotes/docker/pusher.go#L106-L118) but regardless: it's not including all of the schema 2 types), which is resulting in 400s for the client not supporting schema 2. I may look into patching the registry I'm running to stop rejecting this, as I don't believe I have any clients that have used schema 1 in several years, but I was curious if there might now be broader motivation to support Accept: */* that I might expect this to become supported in distribution

**Comment by milosgajdos** [comment]
@thaJeztah, mind weighing in here? or @crazy-max
